### PR TITLE
Implements packageIterator

### PR DESCRIPTION
**Pull request by arcanis**
Offers a fix for the problem described in https://github.com/browserify/resolve/issues/199#issuecomment-570225108

This PR offers a way to alter the default package path generation. The default implementation still uses `path.join('/.../node_modules', packageName)` (which cannot work for PnP when using the link protocol for the reasons described [here](https://github.com/browserify/resolve/issues/199#issuecomment-570393826)), but the PnP resolver (automatically configured thanks to #174) will directly return the path to the package regardless of the dependency type, shortcutting the default logic.

One big advantage of this diff is that it's semver-minor, which avoids having to upgrade all third-party consumers (which I really don't have the bandwidth to do 😅).

### CODE REVIEW

File: lib/async.js [hunk 1]
```diff
@@ -36,6 +36,18 @@ var maybeUnwrapSymlink = function maybeUnwrapSymlink(x, opts, cb) {
...
+var getPackageCandidates = function getPackageCandidates(x, start, opts) {
+    var dirs = nodeModulesPaths(start, opts, x);
+    for (var i = 0; i < dirs.length; i++) {
+        dirs[i] = path.join(dirs[i], x);
+    }
+    return dirs;
+};
+
+var defaultPackageIterator = function defaultPackageIterator(x, start, getPackageCandidates, opts) {
+    return getPackageCandidates();
```
Comment (nicolo-ribaudo) [inline lib/async.js]: Nit: this is really confusing. I saw a function with the same name defined right before this function, which takes 3 arguments. It took me a while to realize that this isn't calling that function, but the one passed as a parameter.

Maybe the one on line 39 could be renamed to `getPackageNodeModulesCandidates`?

Comment (ljharb) [reply lib/async.js]:
```suggestion
    return defaultGetPackageCandidates(x, start, opts);
```

File: lib/async.js [hunk 2]
```diff
@@ -36,6 +36,18 @@ var maybeUnwrapSymlink = function maybeUnwrapSymlink(x, opts, cb) {
...
 
+var getPackageCandidates = function getPackageCandidates(x, start, opts) {
+    var dirs = nodeModulesPaths(start, opts, x);
+    for (var i = 0; i < dirs.length; i++) {
+        dirs[i] = path.join(dirs[i], x);
+    }
+    return dirs;
+};
+
+var defaultPackageIterator = function defaultPackageIterator(x, start, getPackageCandidates, opts) {
```
Comment (ljharb) [inline lib/async.js]:
```suggestion
var defaultPackageIterator = function defaultPackageIterator(x, start, defaultGetPackageCandidates, opts) {
```

File: lib/async.js [hunk 3]
```diff
@@ -261,31 +274,28 @@ module.exports = function resolve(x, options, callback) {
...
-            if (n) return cb(null, n, pkg);
-            processDirs(cb, dirs.slice(1));
+                    iterate(i + 1);
+                });
+            });
         }
     }
     function loadNodeModules(x, start, cb) {
-        processDirs(cb, nodeModulesPaths(start, opts, x));
+        processDirs(cb, packageIterator(x, start, function () {
```
Comment (ljharb) [inline lib/async.js]: This is also confusing. `packageIterator` should be called with the third argument as the default one, ie:
```suggestion
        processDirs(cb, packageIterator(x, start, getPackageCandidates, opts)
```

File: lib/sync.js [hunk 4]
```diff
@@ -38,6 +38,18 @@ var maybeUnwrapSymlink = function maybeUnwrapSymlink(x, opts) {
...
+var getPackageCandidates = function getPackageCandidates(x, start, opts) {
+    var dirs = nodeModulesPaths(start, opts, x);
+    for (var i = 0; i < dirs.length; i++) {
+        dirs[i] = path.join(dirs[i], x);
+    }
+    return dirs;
+};
+
+var defaultPackageIterator = function defaultPackageIterator(x, start, getPackageCandidates, opts) {
+    return getPackageCandidates();
```
Comment (ljharb) [inline lib/sync.js]:
```suggestion
    return getPackageCandidates(x, start, opts);
```

File: lib/sync.js [hunk 5]
```diff
@@ -164,15 +177,16 @@ module.exports = function (x, options) {
     }
 
     function loadNodeModulesSync(x, start) {
-        var dirs = nodeModulesPaths(start, opts, x);
+        var dirs = packageIterator(x, start, function () {
```
Comment (ljharb) [inline lib/sync.js]:
```suggestion
        var dirs = packageIterator(x, start, getPackageCandidates, opts);
```

File: lib/sync.js [hunk 6]
```diff
@@ -164,15 +177,16 @@ module.exports = function (x, options) {
...
+            return getPackageCandidates(x, start, opts);
+        }, opts);
+
         for (var i = 0; i < dirs.length; i++) {
             var dir = dirs[i];
-            if (isDirectory(dir)) {
```
Comment (ljharb) [inline lib/sync.js]: why is this line removed? if you want to pass your own isDirectory that's fine, but this check still needs to be ran.

File: readme.markdown [hunk 7]
```diff
@@ -80,6 +80,12 @@ options are:
     * getNodeModulesDirs - a thunk (no-argument function) that returns the paths using standard `node_modules` resolution
     * opts - the resolution options
 
+* `opts.packageIterator(request, start, opts)` - return the list of candidate paths where the packages sources may be found (probably don't use this)
+    * request - the import specifier being resolved
+    * start - lookup path
+    * getPackageCandidates - a thunk (no-argument function) that returns the paths using standard `node_modules` resolution
```
Comment (ljharb) [inline readme.markdown]: this really does confuse me. a user provides `packageIterator` and their callback receives 4 arguments - why would they receive a thunk here rather than passing through the request and the start?

**Comment by nicolo-ribaudo** [comment]
@arcanis Is `packageIterator` meant to return the possible package folders or the possible resolved path?

i.e. for `resolve("foo/bar")` should it return `"/.../node_modules/foo/bar"` or `"/.../node_modules/foo"`? I am asking because this PR does the first thing, while https://github.com/yarnpkg/berry/blob/3e324d5383de1f4143c0afc4829c33bd1b8e5f50/packages/plugin-compat/extra/resolve/normalize-options.js#L22 does the second one.

File: lib/sync.js (hunk 6 above)
Comment (nicolo-ribaudo) [reply lib/sync.js]: Maybe it could be moved in the default `getPackageCandidates`?

File: readme.markdown (hunk 7 above)
Comment (arcanis) [reply readme.markdown]: My intent was to use the same pattern as `paths`, which takes a thunk here:

https://github.com/browserify/resolve/blob/master/lib/node-modules-paths.js#L35

File: lib/sync.js (hunk 6 above)
Comment (arcanis) [reply lib/sync.js]: Why is the `isDirectory` needed? Is it an optimization? If `dir` isn't a directory then both `loadAsFile` and `loadAsDirectory` will fail at the next `stat` call, so the loop will be roughly the same 🤔

Comment (ljharb) [reply lib/sync.js]: Just like there's a custom `isFile` in v1, there's also a custom `isDirectory` in v2; users are able to overload this logic. If we don't call it, they're unable to do that.

File: readme.markdown (hunk 7 above)
Comment (ljharb) [reply readme.markdown]: hmm, fair point; i made that change in https://github.com/browserify/resolve/pull/173#pullrequestreview-177499139 on top of #172. I suppose we'll stick with it.

File: lib/async.js (hunk 3 above)
Comment (ljharb) [inline lib/async.js]: since the entire `processDirs` and `packageIterator` invocations both can't fit on one line, please format them so that each argument is on a line by itself.
```suggestion
        processDirs(
            cb,
            packageIterator(
              x,
              start,
              function () { return getPackageCandidates(x, start, opts); },
```

File: lib/async.js [hunk 8]
```diff
@@ -261,31 +274,28 @@ module.exports = function resolve(x, options, callback) {
...
+            return getPackageCandidates(x, start, opts);
```
Comment (ljharb) [inline lib/async.js]:
```suggestion
```

File: lib/async.js [hunk 9]
```diff
@@ -261,31 +274,28 @@ module.exports = function resolve(x, options, callback) {
...
+        }, opts));
```
Comment (ljharb) [inline lib/async.js]: (the indentation is probably off here in my suggestion)
```suggestion
            opts
         )
      );
```

File: lib/sync.js (hunk 5 above)
Comment (ljharb) [inline lib/sync.js]: same here wrt formatting

**Comment by arcanis** [comment]
Some context on my last few commits:

- About `isDirectory`, I found out it got added in https://github.com/browserify/resolve/pull/190 for performance reasons. Since it doesn't have semantic meaning, checking the existence of the parent directory is good enough to preserve the original intent, so that's what I did (this avoids having to make `packageIterator` sync or async depending on the caller).

- Regarding @nicolo-ribaudo's question I initially meant to return the package folder only (which isn't the case atm). However, after a closer inspection of the current architecture I feel like this would require to do extra work for little reason. So I think keeping the current behavior is preferable (ie `@scope/foo/bar` will produce `/n_m/@scope/foo/bar`, not `/n_m/@scope/foo`).

- The `isCore` change simply removes an unreachable call in the sync codepath (it could only be reached if the request was both an absolute path and a core module, which isn't possible) and moves the `isCore` check in the async codepath to be done before trying the n_m resolution, similar to what happens in the sync codepath. Due to the way the check was done it doesn't change anything, except that `packageIterator` won't be called if the package is a core module.

File: lib/async.js [hunk 10]
```diff
@@ -261,31 +275,39 @@ module.exports = function resolve(x, options, callback) {
...
+            if (i === dirs.length) return cb(null, undefined);
+            var dir = dirs[i];
 
-        function isdir(err, isdir) {
-            if (err) return cb(err);
-            if (!isdir) return processDirs(cb, dirs.slice(1));
-            var file = path.join(dir, x);
-            loadAsFile(file, opts.package, onfile);
-        }
+            isDirectory(path.dirname(dir), function (err, status) {
```
Comment (ljharb) [inline lib/async.js]: why the extra path.dirname call? shouldn't the `dirs` list already have that done?

Comment (arcanis) [reply lib/async.js]: `dirs` may point to files, for example if you do resolve something like `foo/package.json` then `dirs` will be `['/n_m/foo/package.json']` (that makes `dir` a misnomer, do you want me to rename it into `entry` or similar?).

Comment (ljharb) [reply lib/async.js]: i mean, yes, but in that case `foo` should be the entry that's a directory and `package.json` should be the entry that's not.

Comment (arcanis) [reply lib/async.js]: That's why I call `dirname`: since `dir` is `/n_m/foo/package.json`, the isDir check has to be done against its parent. Even if it was `/n_m/foo` (which would happen if the request was just `foo`), the right behavior would be to check against `/n_m` (since `foo` is allowed to be a file per the Node resolution rules), so the `dirname` would still stand.

**Comment by ljharb** [comment]
This is a lot of change in one PR; would you mind opening up a separate PR with all the bugfixes and refactors, and associated tests, so we can get this semver-minor PR as small as possible?

File: lib/async.js [hunk 11]
```diff
@@ -265,19 +274,18 @@ module.exports = function resolve(x, options, callback) {
         if (dirs.length === 0) return cb(null, undefined);
         var dir = dirs[0];
 
-        isDirectory(dir, isdir);
+        isDirectory(path.dirname(dir), isdir);
```
Comment (ljharb) [inline lib/async.js]: when is the dirname not going to be a directory? the intention here, i believe, is to see if `dir` points to a file or not. It seems like hoisting the `path.join` prevents that from being determined.

Comment (arcanis) [reply lib/async.js]: The `dirname` isn't a directory when it doesn't exist. From what I gather from #190 and #116, this code was only meant to speed up the resolution by checking whether the n_m folder *existed* before doing the file stat calls.

Checking whether the parent of the subpath (ie `/n_m/foo` for `foo/hello`) exists has the same behaviours: if `foo/hello` can be resolved, then `/n_m/foo` will necessarily be a directory.

Comment (arcanis) [reply lib/async.js]: In other words:

| request | partial resolution | isDirectory (before) | isDirectory (after) |
| --- | --- | --- | --- |
| `foo` | `/n_m/foo` | `/n_m` | `/n_m` |
| `foo/bar` | `/n_m/foo/bar` | `/n_m` | `/n_m/foo` |
| `@foo/bar` | `/n_m/@foo/bar` | `/n_m` | `/n_m/@foo` |

The effect will be the same before and after: the two stat calls that the code would have made before #190 (taking the first line as an example, one to check whether `/n_m/foo.js` exists and another to check whether `/n_m/foo/index.js` exists) won't be necessary because `resolve` is able to find out that the directory supposed to contain them doesn't even exist.

Comment (ljharb) [reply lib/async.js]: I see. Do you think that this perf improvement could also be split out into a separate PR (or even just a separate commit prior to your addition)?

Comment (arcanis) [reply lib/async.js]: In this PR it's not a perf improvement, it's necessary to make my PR work with the changes brought by #190 (because with my PR we don't have access to `n_m` paths anymore in `loadNodeModules` - instead we only have `/n_m/<subpath>`). I guess I could revert it and re-apply it, but I'm not sure it would be much clearer 🤔

Comment (ljharb) [reply lib/async.js]: I'm mainly concerned about an unintentional breakage, and my hope would be to release this particular change as a patch prior to releasing this PR as a minor.

Comment (arcanis) [reply lib/async.js]: I think #190 has only been released in the 2.0.0-next.0 branch, so for the purpose of the patch release the `path.dirname` thing can be removed entirely. Do you want me to open a separate PR against the `1.x` branch?

Comment (ljharb) [reply lib/async.js]: #192 was the PR into the 1.x branch.

File: lib/sync.js [hunk 12]
```diff
@@ -38,6 +38,14 @@ var maybeUnwrapSymlink = function maybeUnwrapSymlink(x, opts) {
     return x;
 };
 
+var getPackageCandidates = function getPackageCandidates(x, start, opts) {
+    var dirs = nodeModulesPaths(start, opts, x);
```
Comment (nicolo-ribaudo) [inline lib/sync.js]: Since this function is the same in `sync.js`, maybe it could be extracted to a separate file?

Comment (ljharb) [reply lib/sync.js]: I thought about that, but then it's a separate file that has to be supported for the rest of v1.
